ddtrace/appsec/_iast/_stacktrace.c

```diff
@@ -31,17 +39,41 @@ GET_FILENAME(PyFrameObject* frame)
         return NULL;
     }
     PyObject* filename = PyObject_GetAttrString((PyObject*)code, "co_filename");
-    Py_DecRef(code);
+    Py_DecRef((PyObject*)code);
     return filename;
 }
+
+static inline PyObject* GET_LOCALS(PyFrameObject* frame) {
+    return PyFrame_GetLocals(frame);
+}
+
+static inline PyObject* GET_FUNCTION(PyFrameObject* frame) {
+    PyCodeObject* code = PyFrame_GetCode(frame);
+    if (!code) {
+        return PyUnicode_FromString("");
+    }
+    PyObject* func = PyObject_GetAttrString((PyObject*)code, "co_name");
+    Py_DecRef((PyObject*)code);
+    if (!func) {
+        return PyUnicode_FromString("");
+    }
+    return func;
+}
+
 #else
 #define GET_FRAME(tstate) tstate->frame
 #define GET_PREVIOUS(frame) frame->f_back
-#define GET_FILENAME(frame) frame->f_code->co_filename
+#define GET_FILENAME(frame) ((PyObject*)(frame->f_code->co_filename))
 #define FRAME_DECREF(frame)
 #define FRAME_XDECREF(frame)
 #define FILENAME_DECREF(filename)
 #define FILENAME_XDECREF(filename)
+#define GET_LOCALS(frame) ((PyObject*)(frame->f_locals))
+static inline PyObject* GET_FUNCTION(PyFrameObject* frame) {
+    PyObject* func = frame->f_code->co_name;
+    Py_INCREF(func);
```

Comment:
We are not doing the decref for `func` anywhere, no?

Reply:
Yes, the caller `get_file_and_line` does it for the returned value (`func_name`).